src/string/extglobmatch.c

```diff
@@ -0,0 +1,276 @@
+
+#define cat_all()    do{                                    \
+                          while (--atoms > 0) {             \
+                              glob *b = *--frag;            \
+                              glob *a = *--frag;            \
+                              *frag++ = glob_seq(a,b);           \
+                              free_glob(a);                 \
+                              free_glob(b);                 \
+                          }                                 \
+                    }while(0)
+
+static glob* match_glob = NULL;
+
+static void cleanup(glob **start, glob **end) {
+    while (start != end)
+    {
+        free_glob(*start);
+        ++start;
+    }
+}
+
+static void print_error(char const c) {
+    fprintf(stderr, "pattern error: Unexpected character '%c'\n", c);
+}
+
+static glob* handle_operator(char op, glob const* g)
+{
+    glob* tmp = NULL;
+    switch (op) {
+        case '!': tmp = glob_not(g);
+                  break;
+        case '*': tmp = glob_star(g);
+                  break;
+        case '+': tmp = glob_plus(g);
+                  break;
+        case '@': tmp = glob_at(g);
+                  break;
+        case '?': tmp = glob_opt(g);
+                  break;
+    }
+    return tmp;
+}
+
+#define push_frag(f) *frag++ = f
+#define pop_frag() *--frag
+#define ARRAY_SIZE 100
+
+static glob *compile(char const *pat) {
+    glob *frags[ARRAY_SIZE]={0};
+    glob **frag=frags;
+    struct paren {
+        int alter;
+        int atoms;
+        char op;
+    } paren[ARRAY_SIZE], *pa;
+
+    int alter = 0;
+    int atoms = 0;
+    pa = paren;
+
+    if (!pat) {
+        goto error;
+    }
+
+    for (;;) {
+        unsigned char p = *pat++;
+        if(!p)
+            break;
+        switch(p) {
+            case '!':
+            case '?':
+            case '@':
+            case '*':
+            case '+':
+                if (*pat != '(')
+                    goto character;
+                if (pa >= paren+100) {
+                    fprintf(stderr,"pattern error: to many nested globs.\n");
+                    goto error;
+                }
+                cat_one();
+                *pa = (struct paren) {
+                    .alter = alter,
+                    .atoms = atoms,
+                    .op = p
+                };
+                pa++;
+                alter = 0;
+                atoms = 0;
+                p = *pat++;
+                break;
+
+            case '(':
+                print_error(p);
+                goto error;
+
+            case ')': {
+                    glob *tmp, *a;
+                    if (pa == paren || !atoms) {
+                        print_error(p);
+                        goto error;
+                    }
+                    --pa;
+                    cat_all();
+                    a = pop_frag();
+                    tmp = handle_operator(pa->op, a);
+                    free_glob(a);
+                    if(!tmp){
```

Comment:
This can't reasonably be `NULL`, all pushed operators are among the ones handled by `handle_operator`.